File: src/csharp/Intel/Iced/Intel/DecoderInternal/OpCodeHandlers_Fpu.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if DECODER
using System.Diagnostics;

namespace Iced.Intel.DecoderInternal {
	sealed class OpCodeHandler_ST_STi : OpCodeHandlerModRM {
		readonly Code code;

		public OpCodeHandler_ST_STi(Code code) => this.code = code;

		public override void Decode(Decoder decoder, ref Instruction instruction) {
			Debug.Assert(decoder.state.Encoding == EncodingKind.Legacy);
			instruction.InternalSetCodeNoCheck(code);
			Static.Assert(OpKind.Register == 0 ? 0 : -1);
			//instruction.Op0Kind = OpKind.Register;
			instruction.Op0Register = Register.ST0;
			Static.Assert(OpKind.Register == 0 ? 0 : -1);
			//instruction.Op1Kind = OpKind.Register;
			instruction.Op1Register = Register.ST0 + (int)decoder.state.rm;
		}
	}

	sealed class OpCodeHandler_STi_ST : OpCodeHandlerModRM {
		readonly Code code;

		public OpCodeHandler_STi_ST(Code code) => this.code = code;

		public override void Decode(Decoder decoder, ref Instruction instruction) {
			Debug.Assert(decoder.state.Encoding == EncodingKind.Legacy);
			instruction.InternalSetCodeNoCheck(code);
			Static.Assert(OpKind.Register == 0 ? 0 : -1);
			//instruction.Op0Kind = OpKind.Register;
			instruction.Op0Register = Register.ST0 + (int)decoder.state.rm;
			Static.Assert(OpKind.Register == 0 ? 0 : -1);
			//instruction.Op1Kind = OpKind.Register;
			instruction.Op1Register = Register.ST0;
		}
	}

	sealed class OpCodeHandler_STi : OpCodeHandlerModRM {
		readonly Code code;

		public OpCodeHandler_STi(Code code) => this.code = code;

		public override void Decode(Decoder decoder, ref Instruction instruction) {
			Debug.Assert(decoder.state.Encoding == EncodingKind.Legacy);
			instruction.InternalSetCodeNoCheck(code);
			Static.Assert(OpKind.Register == 0 ? 0 : -1);
			//instruction.Op0Kind = OpKind.Register;
			instruction.Op0Register = Register.ST0 + (int)decoder.state.rm;
		}
	}

	sealed class OpCodeHandler_Mf : OpCodeHandlerModRM {
		readonly Code code16;
		readonly Code code32;

		public OpCodeHandler_Mf(Code code) {
			code16 = code;
			code32 = code;
		}

		public OpCodeHandler_Mf(Code code16, Code code32) {
			this.code16 = code16;
			this.code32 = code32;
		}

		public override void Decode(Decoder decoder, ref Instruction instruction) {
			Debug.Assert(decoder.state.Encoding == EncodingKind.Legacy);
			if (decoder.state.operandSize != OpSize.Size16)
				instruction.InternalSetCodeNoCheck(code32);
			else
				instruction.InternalSetCodeNoCheck(code16);
			Debug.Assert(decoder.state.mod != 3);
			instruction.Op0Kind = OpKind.Memory;
			decoder.ReadOpMem(ref instruction);
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/dec/OpCodeHandlers_Fpu.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.dec;

import com.github.icedland.iced.x86.Instruction;
import com.github.icedland.iced.x86.OpKind;
import com.github.icedland.iced.x86.Register;
import com.github.icedland.iced.x86.internal.dec.OpSize;

final class OpCodeHandler_ST_STi extends OpCodeHandlerModRM {
	private final int code;

	OpCodeHandler_ST_STi(int code) {
		this.code = code;
	}

	@Override
	void decode(Decoder decoder, Instruction instruction) {
		instruction.setCode(code);
		instruction.setOp0Register(Register.ST0);
		instruction.setOp1Register(Register.ST0 + decoder.state_rm);
	}
}

final class OpCodeHandler_STi_ST extends OpCodeHandlerModRM {
	private final int code;

	OpCodeHandler_STi_ST(int code) {
		this.code = code;
	}

	@Override
	void decode(Decoder decoder, Instruction instruction) {
		instruction.setCode(code);
		instruction.setOp0Register(Register.ST0 + decoder.state_rm);
		instruction.setOp1Register(Register.ST0);
	}
}

final class OpCodeHandler_STi extends OpCodeHandlerModRM {
	private final int code;

	OpCodeHandler_STi(int code) {
		this.code = code;
	}

	@Override
	void decode(Decoder decoder, Instruction instruction) {
		instruction.setCode(code);
		instruction.setOp0Register(Register.ST0 + decoder.state_rm);
	}
}

final class OpCodeHandler_Mf extends OpCodeHandlerModRM {
	private final int code16;
	private final int code32;

	OpCodeHandler_Mf(int code) {
		code16 = code;
		code32 = code;
	}

	OpCodeHandler_Mf(int code16, int code32) {
		this.code16 = code16;
		this.code32 = code32;
	}

	@Override
	void decode(Decoder decoder, Instruction instruction) {
		if (decoder.state_operandSize != OpSize.SIZE16)
			instruction.setCode(code32);
		else
			instruction.setCode(code16);
		assert decoder.state_mod != 3 : decoder.state_mod;
		instruction.setOp0Kind(OpKind.MEMORY);
		decoder.readOpMem(instruction);
	}
}
